src/OpenSearch.Client/Mapping/Visitor/PropertyWalker.cs

```diff
@@ -205,11 +205,27 @@ private static Type GetUnderlyingType(Type type)
 			if (type.IsArray)
 				return type.GetElementType();
 
-			if (type.IsGenericType && type.GetGenericArguments().Length == 1
-				&& (type.GetInterfaces().HasAny(t => t == typeof(IEnumerable)) || Nullable.GetUnderlyingType(type) != null))
-				return type.GetGenericArguments()[0];
+			if (ShouldUnwrapType(type))
+			{
+				var returnType = type.GetGenericArguments()[0];
+				if (ShouldUnwrapType(returnType)) // This is needed for types like IEnumerable<int?>.
+				{
+					return returnType.GetGenericArguments()[0];
+				}
+				return returnType;
+			}
 
 			return type;
+
+			bool ShouldUnwrapType(Type ty)
```

Comment:
Is this a pattern elsewhere in the codebase to write a function within another function? Otherwise make this `private static`?

Reply:
Apparently not.
Although, it might be because local functions are a relatively new language feature.
I can change this.

For the test, I saw that there was no tests at all for the `PropertyWalker` and the `PropertiesExtensions` that was using it was marked as internal. I will add some tests now.

Reply:
I guess I also need to update the change-log too?
Anything special I should pay attention to?

Reply:
Nope, keep it short! :) Thanks much.